Enables libmesh preconditioner reuse capability

reverendbedford opened the pull request:
This closes #21868, adding an option for preconditioner reuse to MOOSE.

## Reason
See issue for details, but this can be a good trick to effective decrease the computational cost of problems where the linear solve is a large portion of the run time.

## Design
Adds two new options to the executioner block:
- `reuse_preconditioner` -- enables preconditioner reuse
- `reuse_preconditioner_max_its` -- the heuristic: reuse the same preconditioner until the linear solve requires more iterations than this target.

## Impact
New feature.

reverendbedford committed:
Enables libmesh preconditioner reuse capability

        closes #21868

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1145629/) on 7cdf54f wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/21869/site)

This comment will be updated on new commits.

hugary1995 commented:
I'd like to see an option to force refactorization at the 0th nonlinear iteration at each time step, if not difficult to add.

reverendbedford commented:
I swear the only test that failed when I originally pushed was the documentation...  Can someone please invalidate and rerun the tests?

@hugary1995 uhh, this may be difficult without iterating back to libmesh.  I'm still thinking though, maybe there is a way to do it.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1146456/) on af9af14 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>c2579d</th>
      <th colspan=4>#21869 af9af1</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.11%</td>
      <td>84.11%</td>
      <td>-0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>78650</td>
      <td>78653</td>
      <td>+3</td>
      <td>7</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14854</td>
      <td>14858</td>
      <td>+4</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21869/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21869/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/21869/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/21869/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/21869/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/21869/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/21869/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/21869/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/21869/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/21869/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/21869/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/21869/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/21869/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/21869/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/21869/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/21869/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/21869/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/21869/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/21869/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/21869/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/21869/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/21869/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/21869/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/21869/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/21869/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/21869/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/21869/coverage/xfem)

</details>



This comment will be updated on new commits.

moosebuild commented:
Job [Nonsparse AD](https://civet.inl.gov/job/1146442/) on af9af14 : invalidated by @hugary1995

hugary1995 commented:
I can't do anything about the controlled app tests.

loganharbour commented:
> I can't do anything about the controlled app tests.

you're good. I believe they're all unrelated

bwspenc commented on framework/doc/content/source/systems/NonlinearSystem.md:
````diff
@@ -365,3 +365,70 @@ which there are quite a few nonlinear iterations per timestep, for which the
+  nl_abs_tol = 1e-10
+```
+
+This solver strategy can be very efficient when solving a problem on a 
````
I'd just say 'This solver strategy can be very efficient when the system Jacobian does not changing very much from iteration to iteration, and is especially beneficial when using computationally expensive preconditioners such as those that use direct factorization.'

I think it's a bit misleading/confusing to make comparisons between workstations and clusters.

reverendbedford replied:
I agree, I updated the docs in the commit I'll push in a second.

bwspenc reviewed (changes requested):
A couple of really minor documentation update requests.

Also, our code coverage testing does not give credit for framework code that is only tested in the modules. Would it be possible to add a simple test that uses this (maybe on a diffusion problem) in the framework tests (moose/test/tests)?

GiudGiud commented:
@oanaoana if you have time to look at this one

reverendbedford committed:
Revision to address code review

        Updated docs and added framework test

reverendbedford commented:
@bwspenc Thanks, and I added a framework test

@hugary1995 I'll give it another hour this evening but I think I would need to go back to libmesh to do this (and it would be quite easy if I start over there).  So maybe I'll take this as an IOU, as I'd like to get this merged first.

hugary1995 commented:
Yeah, no worries. That's just a good-to-have option not a must. That additional capability really only shines when you have a lot of nonlinear iterations in one time step. For problems without fixed point iterations, it's not gonna make a difference.

reverendbedford commented:
Yes, I think I'll need to modify libmesh to add the "new preconditioner every time step option."  So I think this is ready for any final reviews and merging.

I don't think the coverage failure is my fault, can someone invalidate?

bwspenc commented:
I think this is good to go, and am ready to merge this. I'm excited to give this a try. If we're ever running completely linear problems, we should be able to get really good performance by setting the reuse max its to infinity. Another potential related option would be to monitor the convergence rate and update the preconditioner when linear convergence deteriorates beyond some threshold.

hugary1995 commented on test/tests/preconditioners/reuse/tests:
```diff
@@ -0,0 +1,20 @@
+    input = convergence.i
+    cli_args = "Executioner/reuse_preconditioner=true Outputs/csv/file_base=reuse_case"
+    csvdiff = reuse_case.csv
+    requirement = 'Preconditioner is reused until the linear iterations exceed '
+                  'the value of reuse_preconditioner_max_its upon which the '
+                  'system recalculates the preconditioner'
```
> Since the example uses NEWTON, the number of linear iterations is just a proxy for number of Newton iterations?

No, the number of linear iterations should be, well... the number of iterations whichever linear solver uses to solve the linear system of equations. In this case the linear solver is gmres IIRC.

reverendbedford replied:
With this option for Newton we are still solving the nonlinear system using the full, "exact" Newton-Raphson method and we're still solving the exact linearized system.  However, our method of solving the linearized system is going to get less and less effective as we reuse a prior preconditioner, hence the heuristic based on the number of linear iterations.  This all works because forming the preconditioner is the most expensive part of this whole process.

For PJFNK preconditioner reuse will work (you can use the same set of options) but I'm not sure how effective it will be.  You're still forming a preconditioner so this has the potential to reduce the amount of work needed to solve the problem but whether or not it's effective is going to depend on the cost of forming the preconditioner versus the cost of iteration.

recuero replied:
> In this case the linear solver is gmres IIRC.

I'll give you that, dumb question. I was thinking direct solve... 

@reverendbedford Thanks for the details: Does the number of linear iterations accumulate across Newton iterations or it will always compute the preconditioner upon starting a new one? (You may have already explained this but I can't see it in the docs)

recuero replied:
Okay, it seems carryover only happens within the Newton solve according to the parameter description.

reverendbedford committed:
Fix typo in NonlinearSystem.md

reverendbedford commented on framework/doc/content/source/systems/NonlinearSystem.md:
```diff
@@ -365,3 +365,66 @@ which there are quite a few nonlinear iterations per timestep, for which the
+
+## Reusing preconditioners id=reuse_preconditioners
+
+The simple version of GRMES and other iterative methods converge only very
```
Ugh, typos.  Thanks.

oanaoana commented:
> @oanaoana if you have time to look at this one

It's a very good idea to include reusability. My only comment is more to a PETSc lack of clarity that propagates. Using full LU as a preconditioner is overkill since it already solves the problem, and most likely there's some incomplete LU used for -pc_type LU. In the context of reusability though it becomes interesting cause one could do a full LU at one step, and as the simulation advances this LU becomes inexact and a true preconditioner.

GiudGiud commented on framework/src/executioners/FEProblemSolve.C:
```diff
@@ -126,11 +126,25 @@ FEProblemSolve::validParams()
+                        "If true reuse the previously calculated "
+                        "preconditioner for the linearized "
+                        "system across multiple solvers, "
+                        "controlled by reuse_preconditioner_max_its");
```
```suggestion
                        "within nonlinear iterations in a timestep and "
                        "across timesteps. The preconditioner is only reset as "
                        "controlled by reuse_preconditioner_max_its");
```

GiudGiud commented on test/tests/preconditioners/reuse/convergence.i:
```diff
@@ -0,0 +1,100 @@
+# Simple 3D test with diffusion, setup to make sure
+# there is a sensible difference in the linear iteration
+# counts with resuse versus without resuse
```
```suggestion
# counts with re-use versus without re-use
```

GiudGiud commented on framework/doc/content/source/systems/NonlinearSystem.md:
```diff
@@ -365,3 +365,66 @@ which there are quite a few nonlinear iterations per timestep, for which the
+for a large simulation is numerically expensive.
+
+One option is to reuse a preconditioner for many iterations of the 
+linear iterative solver.  The preconditioner can be carried over
```
```suggestion
nonlinear iterative solver.  The preconditioner can be carried over
```

GiudGiud replied:
this might not be the clearest way to disambiguate

reverendbedford replied:
You're reusing the linear preconditioner so it seems correct as written.  But I rephrased it to try to make it more clear.

GiudGiud commented on framework/doc/content/source/systems/NonlinearSystem.md:
```diff
@@ -365,3 +365,66 @@ which there are quite a few nonlinear iterations per timestep, for which the
+in the `[Executioner]` block will reuse the same preconditioner until
+the number of linear iterations required to solve the linearized system of
+equations exceeds 20.  At that time the solution algorithm will reform
+the preconditioner and repeat the process.
```
```suggestion
the preconditioner and repeat the process.
```
make it clear whether it:
- in the current nonlinear iteration, accepts the result for this iteration, makes a new PC then does the next NL iteration
or
- within the nonlinear iteration, makes a new PC then tries again to reach convergence

reverendbedford replied:
It will keep going until it either solves the current system or reaches `l_max_its`.  It will then form a new preconditioner in the next nonlinear iteration.

Updated the docs.

GiudGiud commented on framework/doc/content/source/systems/NonlinearSystem.md:
```diff
@@ -365,3 +365,66 @@ which there are quite a few nonlinear iterations per timestep, for which the
+
+  reuse_preconditioner = true
+  reuse_preconditioner_max_its = 20
+
+  nl_max_its = 10
+  nl_rel_tol = 1e-8
+  nl_abs_tol = 1e-10
```
```suggestion
```
I d rather users dont set this. Pretty often this doesnt go well

GiudGiud commented on framework/doc/content/source/systems/NonlinearSystem.md:
````diff
@@ -365,3 +365,66 @@ which there are quite a few nonlinear iterations per timestep, for which the
+```
+  petsc_options_iname = '-pc_type -pc_factor_mat_solver_package -ksp_type'
+  petsc_options_value = 'lu superlu_dist gmres'
+  l_tol = 1e-8
+  l_max_its = 100
````
```suggestion
```

GiudGiud reviewed (changes requested):
I dont see any logic to error out or force a recomputation when the number of elements / dofs changes because of adaptivity.
Is this in libmesh? Could you please point me to it?

The interface to trigger or stop re-using is pretty limited. I think this is ok for where we are currently though, but eventually we would want to be to specify it on particular timesteps etc. Maybe even leverage the EXECUTE_ON flags

reverendbedford commented on framework/doc/content/source/systems/NonlinearSystem.md:
```diff
@@ -365,3 +365,66 @@ which there are quite a few nonlinear iterations per timestep, for which the
+
+One option is to reuse a preconditioner for many iterations of the 
+linear iterative solver.  The preconditioner can be carried over
+through nonlinear iterations and even across load steps.  MOOSE
```
Yes.

reverendbedford commented:
So everything @GiudGiud suggested is easy to fix except the valid point about adaptive mesh refinement.  Right now this will not work because with `preconditioner_reuse = true` it will just go ahead and try to reuse the same, incorrectly-sized preconditioner on the new system.  It looks like I need to circle back to libmesh and add a "new preconditioner right now" option, which I can then also use to implement @hugary1995 request...

hugary1995 commented:
> So everything @GiudGiud suggested is easy to fix except the valid point about adaptive mesh refinement. Right now this will not work because with `preconditioner_reuse = true` it will just go ahead and try to reuse the same, incorrectly-sized preconditioner on the new system. It looks like I need to circle back to libmesh and add a "new preconditioner right now" option, which I can then also use to implement @hugary1995 request...

This will also solve the issue I mentioned to you before about the variational inequality solvers which also resizes the preconditioner across nonlinear iterations.

hugary1995 commented:
So I guess it's beneficial to circle back to libmesh for various reasons :-)

reverendbedford committed:
Fix preconditioner use with AMR

        This tests a libmesh change to fix the problem behavior with
        AMR (where petsc would doggedly try to reuse a preconditioner
        with the wrong size)

reverendbedford commented:
@GiudGiud and @hugary1995 this commit addresses your issues.  But it won't run until [this](https://github.com/libMesh/libmesh/pull/3383) makes its way into libmesh and then over into MOOSE...

reverendbedford commented:
I gave up on @hugary1995 's issue with the variational inequality solver.  That will require a change over in petsc, so I'll gripe about it to them.  Once the libmesh change makes its way over this will fix the AMR issue though and would probably be worth merging in.

lindsayad commented on framework/doc/content/source/systems/NonlinearSystem.md:
```diff
@@ -365,3 +365,64 @@ which there are quite a few nonlinear iterations per timestep, for which the
+
+A great number of preconditioners exist, but 
+[multigrid](https://en.wikipedia.org/wiki/Multigrid_method)
+methods are often among the best choices.  
```
```suggestion
methods are often among the best choices for problems without significant hyperbolic character.
```

lindsayad commented on framework/doc/content/source/systems/NonlinearSystem.md:
```diff
@@ -365,3 +365,64 @@ which there are quite a few nonlinear iterations per timestep, for which the
+
+Using these parameters in combination with a direct factorization of the
+system can be very efficient.  The following is an example of how to
+configure PETSc and MOOSE to solve the equations with this combination:
```
```suggestion
direct PETSc and MOOSE to solve the equations with this combination:
```

reverendbedford replied:
I suppose configure is a loaded term

lindsayad commented on framework/doc/content/source/systems/NonlinearSystem.md:
```diff
@@ -365,3 +365,64 @@ which there are quite a few nonlinear iterations per timestep, for which the
+in the `[Executioner]` block will reuse the same preconditioner until
+the number of linear iterations required to solve the linearized system of
+equations exceeds 20.   If the number of linear iterations exceeds 
+`reuse_preconditioner_max_its`
```
I missed the boat on the libMesh PR, but I would vote to rename this parameter to `reuse_preconditioner_max_linear_its`. I know that we can set preconditioner evaluation every x number of nonlinear iterations via the PETSc command line option, but maybe in the future we would want a shim option in MOOSE like `reuse_preconditioner_nl_its` (or with a max in there) and then having the distinction between linear-based and nonlinear based would be really nice. Even if we don't ever add any nonlinear based MOOSE Executioner parameters, being more verbose with the the parameter name is hardly a bad thing in my opinion.

reverendbedford replied:
Yes, I suppose this naming makes more sense.  As we have to change things over there to get AMR to work anyway I'll go ahead and make the change.

hugary1995 replied:
FYI there's a new sailing boat https://github.com/libMesh/libmesh/pull/3383

lindsayad reviewed (changes requested):
I don't love diffs that involve iteration counts as those may vary from PETSc version to PETSc version. Is there another way we can test this without having iteration counts in diffs?

reverendbedford commented:
@lindsayad I'm open to suggestions but I think we're doomed to do it this way.  The heuristic is after all "refactor when you exceed some specific number of linear iterations."  The idea behind the tests is you can tell that it correctly triggered by examining the linear iteration count -- it goes back to 1 when we get a new preconditioner.

reverendbedford committed:
Code review fixes

        - Updates the name of the parameter to
          `reuse_preconditioner_max_linear_its`
        - Various doc fixes and changes

reverendbedford commented:
Thanks @hugary1995, yes, this will now require https://github.com/libMesh/libmesh/pull/3383 to be merged in first.  And the tests won't pass until then.

lindsayad commented:
Ok. At least it passes two PETSc versions right now. Can revisit if there's ever a problem switching PETSc versions

reverendbedford commented:
@lindsayad if you trigger a build this should now work as it looks like the libmesh changes made their way in.

moosebuild commented:
All [jobs](https://civet.inl.gov/event/97524/) on 7f8b072 : invalidated by @GiudGiud

post-libmesh update

lindsayad commented:
we will need to wait for next to get un-snarled, and then this looks like it's ready to go

moosebuild commented:
Job [External app tests](https://civet.inl.gov/job/1162414/) on 7f8b072 : invalidated by @lindsayad

lindsayad commented:
@GiudGiud has your review been addressed?

GiudGiud reviewed (dismissed):
awesome

we should document the difference between this and passing petsc these two:
```
	-snes_lag_preconditioner <lag>	- how often preconditioner is rebuilt (use -1 to never rebuild)
-snes_lag_jacobian <lag>	- how often Jacobian is rebuilt (use -1 to never rebuild)
```

reverendbedford commented:
@GiudGiud 

`-snes_lag_preconditioner` versus `reuse_preconditioner`:
1. The old libmesh behavior (and new default behavior if you don't set `reuse_preconditioner=true`) was the throw away the SNES instance between time steps, meaning you couldn't retain the preconditioner across time steps.
2. There is also a difference in behavior: `-snes_lag_preconditioner` recalculates the preconditioner every X linear iterations, `reuse_preconditioner` reuses the preconditioner until you hit `reuse_preconditioner_max_its`.  The latter behavior needs to tack on a PETSc monitor to make it work.

`lag_preconditioner` versus `lag_jacobian`:
- Preconditioner lag affects convergence in the linear iterations but not convergence in the nonlinear iterations.  Conversely, it potentially speeds up the linear solve and does not affect the nonlinear solve.
- Jacobian reuse affects convergence in the nonlinear iterations but not convergence in the linear iterations.  Conversely, it potentially speeds up the nonlinear solve but does not affect the linear solve.

Does that make sense?  I don't know how clear either explanation is.

lindsayad commented:
> * Preconditioner lag affects convergence in the linear iterations but not convergence in the nonlinear iterations.  Conversely, it potentially speeds up the linear solve and does not affect the nonlinear solve.

This assumes that you are able to solve the linear system to the same tolerance (and importantly to the same true residual norm monitored with -ksp_monitor_true_residual). If not, then your nonlinear iteration history will likely be different

lindsayad reviewed (changes requested):
@reverendbedford sorry to tack on one final request, but per @GiudGiud's suggestion can you add your explanation about the difference between your option and `-snes_lag_xxx` in the `NonlinearSystem.md` documentation?

reverendbedford commented:
That's true, though you can play with  `l_max_its` to try to make sure that you always achieve your `l_tol`.  Put another way, the linear solver doesn't immediately give up once it hits `reuse_preconditioner_max_linear_its`, it will keep trying out to `l_max_its` and only next nonlinear iteration make a new preconditioner.  But yes, you can run into situations where the old preconditioner is really bad and you exceed `l_max_its` without hitting your tolerance.

Let me add this to the docs

reverendbedford committed:
Documented reuse option differences

Documented the differences between the various Jacobian and
preconditioner reuse options available in PETSc and the new options
made available with this PR.

lindsayad commented on framework/doc/content/source/systems/NonlinearSystem.md:
```diff
@@ -427,3 +427,44 @@ when the cost of solving the linearized system is a large fraction of
+
+There are two differences between
+`reuse_preconditioner` and 
+`reuse_preconditioner_max_linear_its`
```
```suggestion
```

lindsayad commented on framework/doc/content/source/systems/NonlinearSystem.md:
```diff
@@ -427,3 +427,44 @@ when the cost of solving the linearized system is a large fraction of
+There are two differences between
+`reuse_preconditioner` and 
+`reuse_preconditioner_max_linear_its`
+and setting up preconditioner reuse directly in PETSc with the 
```
```suggestion
setting up preconditioner reuse directly in PETSc with the 
```

lindsayad commented on framework/doc/content/source/systems/NonlinearSystem.md:
```diff
@@ -427,3 +427,44 @@ when the cost of solving the linearized system is a large fraction of
+`reuse_preconditioner` and 
+`reuse_preconditioner_max_linear_its`
+and setting up preconditioner reuse directly in PETSc with the 
+`snes_lag_preconditioner_persists` and `-snes_lag_preconditioner` options:
```
```suggestion
`-snes_lag_preconditioner_persists` and `-snes_lag_preconditioner` options:
```

reverendbedford committed:
Edits

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

reverendbedford committed:
Edits

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

reverendbedford committed:
Edits

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

GiudGiud commented:
Could you please squash the (adjacent) documentation commits together?
This is ready to merge once commit history is clean

GiudGiud commented:
And thanks for the additional details! They all make sense to me. This PR definitely adds a more tuned control of the PC reuse than flat numbers
